docs/integrations/custom_middleware.md

```diff
@@ -0,0 +1,191 @@
+    stream: BaseStream, fn: Callable, context_manager: ContextManagerType | None
+):
+    # handle after stream here
+```
+
+The first argument will be a Mirascope `Stream` of the provider you are using.
+See `handle_call_response` section for information regarding the other arguments.
+
+One thing to note for `handle_stream` is that it will not be called until after the `Generator` has been exhausted.
+
+### `handle_response_model` and `handle_response_model_async`
+
+These functions will be called after making a mirascope call with `response_model` set with the following signature:
+
+```python
+from typing import Callable
+from pydantic import BaseModel
+
+from mirascope.core.base._utils._base_type import BaseType
+
+def handle_response_model(
+    response_model: BaseModel | BaseType, fn: Callable, context_manager: ContextManagerType | None
+):
+    # handle after call with response_model here
+```
+
+The first argument will be a Pydantic `BaseModel` or Python primative depending on what type `response_model` is.
+See `handle_call_response` section for information regarding the other arguments.
+
+For `BaseModel` you can grab the `CallResponse` via `response_model._response`.
+However for primatives `BaseType`, this information is not available.
+
+### `handle_structured_stream` and `handle_structured_stream_async`
+
+These functions will be called after streaming a mirascope call with `response_model` set with the following signature:
+
+```python
+from typing import Callable
+
+from mirascope.core.base._structured_stream import BaseStructuredStream
+
+def handle_structured_stream(
+    result: BaseStructuredStream,
+    fn: Callable,
+    context_manager: ContextManagerType | None,
+):
+```
+
+The first argument will be a Mirascope `StructuredStream` of the provider you are using.
+Like with `handle_stream`, `handle_structured_stream` will not be called until the `Generator` has been exhausted.
+
+### `custom_context_manager`
+
+You can define your own context manager where the yielded object is passed to each of the handlers.
+
+Here is an example using [SQLModel](https://sqlmodel.tiangolo.com/):
+
+```python
+from contextlib import contextmanager
+from typing import Any, Callable, Generator
+
+from sqlmodel import Session
+
+from mirascope.core.base import BaseCallResponse
+
+
+@contextmanager
+def custom_context_manager(
+    fn: Callable,
+) -> Generator[Session, Any, None]:
+    engine = ...
+    with Session(engine) as session:
+        yield session
+
+def handle_call_response(
+    result: BaseCallResponse, fn: Callable, session: Session | None
+):
+    # write to db 
+```
+
+In this example, the 3rd argument of your handler will be typed with the yielded `Session` and metadata can be written to your database of choice.
+
+### `custom_decorator`
+
+There may be existing libraries that have a decorator implemented already. You can pass that decorator in to `custom_decorator` which will wrap the Mirascope call with that decorator.
```

Comment:
this can be more clear around what's happening i.e. in what order is the decorator called wrt. the middleware and other handlers (before or after?)

Reply:
Okay so just mention this decorator is called before?